write empty tseries with a single NA value

jasonjensen opened the pull request:
This is a proposed fix for Issue https://github.com/bankofcanada/FAME.jl/issues/8

Empty TSeries are written with a single NA value in the first period of their range. Series with a single NA value are in turn interpreted as an empty TSeries when being written.

There are also some tests added.

jasonjensen committed:
write empty tseries with a single NA value

jasonjensen committed:
fix treatment of empty boolean series

bbejanov commented on src/Bridge.jl:
```diff
@@ -556,6 +573,10 @@ function _unmissing!(x::Ref, v::Val{T}) where {T}
 function _unmissing!(x::Vector, v::Val{T}) where {T}
+    # a single NaN value is an empty TSeries
+    if length(x) == 1 && _ismissing(x[1], v)
+        return typeof(x)()
```
I suggest `return empty!(x)` -- mutate the existing `x` according to the trailing ! convention. 

Although, looking at how `_unmissing!()` is actually used, your `return typeof(x)()` will also work correctly.

bbejanov commented on test/runtests.jl:
```diff
@@ -68,3 +68,30 @@ end
+        t3 = TSeries(1996Q2, Vector{Bool}()),
+        t4 = TSeries(1996Q2, Vector{Bool}([true, false, true]))
+    )
+    writefame("data.db", w)
```
You can also use the work database for the tests, instead of writing a file on disk. `db = workdb()` gets you a handle, which you can pass instead of filename.  Just a suggestion. The only advantage is that using the work db you don't have to delete data.db after.

bbejanov replied:
Btw, `TSeries` has a constructor where the eltype is the first argument. A bit more compact. e.g. `TSeries(Bool, 1996Q2)`.

bbejanov commented on src/Bridge.jl:
```diff
@@ -507,6 +514,16 @@ function refame(name::Symbol, value::TSeries)
+            val = [FNUMNA]
+        elseif ElType == Bool
+            val = [FBOONA]
+        elseif ElType == Float64 || !(ElType <: MIT)
```
I don't understand why we have `!(ElType <: MIT)` here and we are missing the `elseif ElType <: MIT` branch, in which I think we need `push!(val,  FAME_INDEX_NA)`.  You can try a test with a TSeries of eltype MIT, something like `TSeries(MIT{Monthly}, 1995Q1))`, to see how it might work.

Also (and this is very nitpicky, but please hear me out) in this if-block I would use `push!(val, F<XYZ>NA)` to mutate the existing `val` and also I would start with `if isempty(val)` rather than `if isempty(value)`. At this point in the code `isempty(value) == isempty(val)`, so practically either one would work. But conceptually, at this point in the code we have already converted `value` to `val` and we just want to fix `val` for the case of it being empty. So, why check `values` if `val` is all we care about? Also, I'm worried about creating a Vector with plain `[` where the compiler has to infer the eltype from the elements. We are very careful above to create `val` with the correct eltype being explicitly specified; we need such tight control because `val` will be passed to the C library and if it ends up being the wrong type, we might corrupt the database. So, I am more comfortable using `push!` here, which is guaranteed to preserve the eltype.

bbejanov commented on src/Bridge.jl:
```diff
@@ -481,8 +488,8 @@ end
+        lind = find
     else
         (fr, find) = _mit_to_date(firstdate(value))
         lind = find + length(value) - 1
```
I think this entire if-statement is equivalent to just running the code in its else-branch.

bbejanov replied:
> I think this entire if-statement is equivalent to just running the code in its else-branch.

No actually it's not. But I still think we should get rid of the if-statement and keep only the `(fr, find) = _mit_to_date(firstdate(value))` here. We should compute `lind` at the very end, just before we call the `FameObject` constructor and after we "fix" `val`.  Then we can do `lind = find + length(val) - 1` to be sure that we will write as many values as we have stored in the FameObject.

bbejanov reviewed (approved):
Awesome, thanks Jason for taking care of this! 

My comments are all cosmetic, except for one thing - I think it will not add a missing value in 
 an empty time series where the values type is <:MIT.  Also, the fix doesn't seem to be implemented for time series of eltype string, but this is a corner case that's probably not worth the effort.

bbejanov reviewed (changes requested):
Actually, I think we should fix the calculation of `lind` in `refame` to be sure that it reflects the true length of `val`. The rest of my comments are not binding.

jasonjensen committed:
fix empty MIT series, improve treatment of other empty series

jasonjensen commented:
Thank you for your thurough review @bbejanov , I really appreciate it.

I have addressed your issues and incorporated your suggestions. W.r.t. the TSeries of MITs, I explored implementing FAME's missing DATE values (i.e. FDATNA) but that did not seem to work with the MIT values.

I also tried using the work_db() in the tests but it turns out that it was affected by the other tests so I kept it to a written db to ensure the test would not interfere with each other.

bbejanov commented on src/Bridge.jl:
```diff
@@ -261,12 +261,24 @@ unfame(fo::FameObject{:series,:precision,FR}) where {FR} =
 unfame(fo::FameObject{:series,:string,FR}) where {FR} =
     copy(fo.data) # error("Can't handle string series")
+function unfame(fo::FameObject{:series,TY,FR}) where {TY,FR}
+    if length(fo.data) == 1 && _ismissing(fo.data[1], Val(:FR))
```
I think it should be `Val(FR)` not `VAL(:FR)`

jasonjensen replied:
Thanks for catching that. It turns out it was still working because the value doesn't get used, it just gets routed to the date-checking `_ismissing` which accepts all the values that aren't in the other `_ismissing` signatures.

I changed it to `Val(TY)` since we're technically checking against the data type in the TSeries, which can be different in than the frequency of the series itself. Neither value gets used, but it's still good to be consistent.

bbejanov replied:
Indeed, you're right!

bbejanov reviewed (approved):
Looks great, thank Jason. You have a little typo (I think) in `unfame()` on line 267. 

You can reset the work database by calling `init_chli()` - it unloads and reloads the FAME library - so it would clean up anything that may be left over from previous tests.

About the missing dates constants, it's not clear from the CHLI help, but looking into hli.h it seems to change - FDATNA/ND/NC (32 bit) in version 9.x and FAME_INDEX_NA/ND/NC (64bit) in 10.x.  The code we have in FAME.jl so far assumes that the version of FAME is at least 10. If we want to become backwards compatible with 9.x, we can do that in a separate PR. For now, I think we can simply use FAME_INDEX_NA. 

![image](https://user-images.githubusercontent.com/12414774/203638372-93e48bcb-3a96-427c-9359-f8d9e14fb3db.png)

Awesome, thanks so much for fixing this!

jasonjensen committed:
use workdb in tests

jasonjensen committed:
fix typo

bbejanov commented:
Feel free to merge when you're ready - looks great to me. Thanks again @jasonjensen !
